### DISCUSSION TITLE
Gap Distance Calculation Issue.

### DISCUSSION THREAD

**Discussion by lucuo** [idaholab/moose#29059] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

<Your question here>Hello, Moose developers. I am performing heat transfer and mechanical calculations for two 3D blocks, with coupled gap distance heat transfer calculation set on two surfaces. These two surfaces have a certain curvature, and the initial gap distance is equal (the two surfaces can be viewed as arcs in the xy-plane, stretched in the z-direction). 
The initial gap distance is correct; however, when one block undergoes displacement in the z-direction, the gap distance calculation produces errors at certain locations, with the gap nearly exceeding the distance limit. The positions where this occurs are distributed in a ring-like pattern, perpendicular to the z-axis.
I used `NearestNodeDistanceAux` to output the gap distance between the two surfaces, but this problem still exists.
Here are my mesh Settings
```
  patch_size = 150
  patch_update_strategy = iteration
  partitioner = centroid 
  centroid_partitioner_direction = z
```
and ThermalContact
```
  quadrature = true
  gap_geometry_type = CYLINDER
  cylinder_axis_point_1 = '0 0 0'
  cylinder_axis_point_2 = '0 0 1'
```
and Executioner
```
  type = Transient
  solve_type = 'PJFNK'
  petsc_options_iname = '-ksp_gmres_restart -pc_type -pc_hypre_type -pc_hypre_boomeramg_max_iter'
  petsc_options_value = '201    hypre   boomeramg    4'
  line_search = 'none'
  automatic_scaling = true
  start_time = 0.0
  end_time = 10
  dtmax = 1
```
and contact
```
    primary = 1
    secondary = 2
    penalty = 1e8
    formulation = PENALTY
    model = frictionless
```
Is there any mistake in my setup or something that needs improvement?

**Comment by GiudGiud** [comment]
Hello

Do you have a figure to show what is happening? I m having trouble conceptualizing it

Guillaume

**Reply by lucuo** [reply]
I use NearestNodeDistanceAux to output the distance between the two surfaces, which is a cloud map of the distance between one surface (with paraview scaling 10:1 in the z direction).
There is no problem in the initial calculation, but once a surface is displaced in the z direction, the calculation becomes problematic
![1](https://github.com/user-attachments/assets/7646e97a-e56d-47c1-b8ba-5d2610f3df58)

At the same time, when I use thermalcontact to calculate the heat transfer between the two surfaces, I also have problems because of the wrong gap distance.

**Reply by GiudGiud** [reply]
in the NearestNodeAux, did you set use_displaced_mesh = true ?

> At the same time, when I use thermalcontact to calculate the heat transfer between the two surfaces, I also have problems because of the wrong gap distance

Is this one computing its own distance?
What syntax are you using? (block as well as parameters)

**Reply by lucuo** [reply]
```
NearestNodeDistanceAux::validParams()
{
  InputParameters params = AuxKernel::validParams();
  params.addClassDescription(
      "Stores the distance between a block and boundary or between two boundaries.");
  params.addRequiredParam<BoundaryName>("paired_boundary", "The boundary to find the distance to.");
  params.set<bool>("use_displaced_mesh") = true;
  return params;
}
```
The default setting for `NearestNodeDistanceAux` is `use_displaced_mesh = true`, which I used

**Reply by lucuo** [reply]
`ThermalContactAction.C `has `params.set<std::vector<VariableName>>("gap_distance") = {"penetration"}; `I use gap_distance to calculate the gap heat transfer coefficient and update the heat transfer coefficient.

**Reply by GiudGiud** [reply]
> params.set<std::vector<VariableName>>("gap_distance") = {"penetration"}; 

This means we use the penetration auxvariable to compute the gap distance. 
You should be able to output it the same way you output the nearest-node distance (is that one used in the calculation?)

Are you running this in parallel?
Any reason you are splitting the domain in Z using the centroid partitioner?
